docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,363 @@
+
+        folder = promiser
+        url = attributes["repo"]
+
+        if os.path.exists(folder):
+            self.promise_kept()
+            return
+
+        self.log_info(f"Cloning '{url}' -> '{folder}'...")
+        os.system(f"git clone {url} {folder} 2>/dev/null")
+
+        if os.path.exists(folder):
+            self.log_info(f"Successfully cloned '{url}' -> '{folder}'")
+            self.promise_repaired()
+        else:
+            self.log_error(f"Failed to clone '{url}' -> '{folder}'")
+            self.promise_not_kept()
+
+
+if __name__ == "__main__":
+    GitPromiseTypeModule().start()
+```
+
+### Logging
+
+Log levels in CFEngine are explained here:
+https://github.com/cfengine/core/blob/master/CONTRIBUTING.md#log-levels
+
+In short, when writing a promise module, these log levels should be used:
+
+* `critical` - Serious errors in protocol or module itself (not in policy)
+* `error` - Errors when validating / evaluating a promise, including syntax errors and promise not kept
+* `info` - Changes made (repaired promises)
+* `verbose` - Human understandable detailed information about promise evaluation
+* `debug` - Programmer-level information that is only useful for CFEngine developers or module developers
+
+### Results
+
+Each operation performed by the module, sends a result back to the agent.
+The possible results are as follows:
+
+* Shared between operations:
+  * `error` - an unexpected error occured in the module or protocol, indicating a bug in CFEngine or the promise module
+    * Should be explained by a `critical` level log message
+* Promise validation:
+  * `valid` - No problems with the data or data types in promise
+  * `invalid` - There are problems with the promise data or data types
+    * Should be explained by an `error` level log message
+* Promise evaluation:
+  * `kept` - promise satisfied already, no change made
+  * `repaired` - promise not satisfied before, but fixed now
+    * The change should be explained in a `info` level log message
+  * `not_kept` - promse not satisfied before, and could not be fixed
+    * Should be explained by an `error` level log message
+* Teminate:
+  * `success` - Module succesfully terminated without errors
+  * `failure` - There were problems when trying to clean up / terminate
+    * Should be explained by a `critical` level log message
+
+Built-in CFEngine promises may have multiple outcomes when evaluated.
+For the sake of simplicity, custom promises may only have 1 outcome, logic should be as follows:
+
+1. If the promise was already satisfied and no changes were necessary, the promise was `kept`
+2. Otherwise, if there were issues and the module could not fix/satisfy all parts of the promise, it was `not_kept`.
+3. Otherwise, if some changes were made (successfully), the promise was `repaired`.
+
+### Protocol
+
+To implement a promise module from scratch, you will have to implement the promise module protocol.
+2 variants of the protocol is provided.
```

Comment:
s/is/are/ right? The subject being "variants", plural.

Reply:
Thanks, fixed.